File: android_pos_app/app/src/main/java/com/pos2013/offline/utils/Protocol2013Helper.kt
package com.pos2013.offline.utils

import android.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object Protocol2013Helper {

    fun generateStan(lastStan: Int): String {
        val nextStan = (lastStan + 1) % 1_000_000
        return String.format("%06d", nextStan)
    }

    fun generateSignature(
        protocolVersion: String,
        merchantId: String,
        terminalId: String,
        batchId: String,
        timestamp: Long,
        nonce: String,
        transactionCount: Int,
        secretKey: String
    ): String {
        val data = "$protocolVersion|$merchantId|$terminalId|$batchId|$timestamp|$nonce|$transactionCount"
        val hmacSha256 = "HmacSHA256"
        val secretKeySpec = SecretKeySpec(secretKey.toByteArray(), hmacSha256)
        val mac = Mac.getInstance(hmacSha256)
        mac.init(secretKeySpec)
        val signatureBytes = mac.doFinal(data.toByteArray())
        return Base64.encodeToString(signatureBytes, Base64.NO_WRAP)
    }
}

File: backend_csharp_api/Utils/Protocol2013Helper.cs
namespace Pos2013.Api.Utils;

public static class Protocol2013Helper
{
    // ISO-8583 requires STAN to be 6 digits (000001 - 999999) and wrap around
    public static string GenerateStan(int lastStan)
    {
        int nextStan = (lastStan + 1) % 1000000;
        return nextStan.ToString("D6");
    }

    // Matches Android's Protocol2013Helper.buildSignatureString
    public static string BuildSignatureString(
        string protocolVersion,
        string merchantId,
        string terminalId,
        string batchId,
        string timestamp,
        string nonce,
        int transactionCount)
    {
        // protocolVersion|merchantId|terminalId|batchId|timestamp|nonce|transactionCount
        return $"{protocolVersion}|{merchantId}|{terminalId}|{batchId}|{timestamp}|{nonce}|{transactionCount}";
    }
}
